blockprod/src/detail/block_maker.rs

```diff
@@ -92,20 +93,66 @@ impl BlockMaker {
         Ok(returned_accumulator)
     }
 
-    pub fn make_block(
+    pub async fn make_block(
         &self,
         current_tip_id: Id<Block>,
-        accumulator: &dyn TransactionAccumulator,
+        transactions: Transactions,
     ) -> Result<Block, BlockProductionError> {
         // TODO: this isn't efficient. We have to create the header first, then see if it obeys consensus rules, then construct the full block
         let current_time = self.time_getter.get_time();
-        let block = Block::new(
-            accumulator.transactions().clone(),
+        let mut block = Block::new(
```

Comment:
This can be made more efficient relatively easily, which you can make in another PR if you wish. Instead of copying everything into this block every time, you can create a new struct, say `BlockTemplate` (?), which has references to all the components. Since it's using references, you won't need to copy anything into it (except for little things that don't matter). Then, once you're done mining it, you have a method in it called `into_block(self)`, which will consume it and create a block. This way you'll copy the transactions only once (and any other very expensive part).

Reply:
Yep, sounds good.

Reply:
I'm not sure this is needed anymore. As `finalize_consensus_data()` has been added into `make_block()`, the result of the function should be a finalized block. So we shouldn't be need to try again any longer so any copying will be done only once.